File: apps/ios/self-management/Core/Services/TokenStorage.swift
//
//  TokenStorage.swift
//  self-management
//
//  CORE SERVICE: Local Token Persistence
//

import Foundation

/// Service responsible for persisting authentication tokens locally.
/// Uses UserDefaults for simplicity in a learning project.
/// NOTE: For production apps, use Keychain for secure token storage.
nonisolated
final class TokenStorage: Sendable {
    
    // MARK: - Constants
    
    private enum Keys {
        static let accessToken = "auth_access_token"
        static let refreshToken = "auth_refresh_token"
    }
    
    // MARK: - Dependencies
    
    private let defaults: UserDefaults
    
    // MARK: - Initialization
    
    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }
    
    // MARK: - Save
    
    /// Save authentication tokens locally after a successful login
    /// - Parameter tokens: The access and refresh tokens to persist
    func saveTokens(_ tokens: AuthTokens) {
        defaults.set(tokens.accessToken, forKey: Keys.accessToken)
        defaults.set(tokens.refreshToken, forKey: Keys.refreshToken)
    }
    
    // MARK: - Load
    
    /// Load previously saved authentication tokens from local storage
    /// - Returns: The stored AuthTokens if available, nil otherwise
    func loadTokens() -> AuthTokens? {
        guard let accessToken = defaults.string(forKey: Keys.accessToken),
              let refreshToken = defaults.string(forKey: Keys.refreshToken)
        else {
            return nil
        }
        
        return AuthTokens(accessToken: accessToken, refreshToken: refreshToken)
    }
    
    // MARK: - Clear
    
    /// Remove all stored tokens (on logout)
    func clearTokens() {
        defaults.removeObject(forKey: Keys.accessToken)
        defaults.removeObject(forKey: Keys.refreshToken)
    }
}

File: apps/mobile/lib/core/network/token_storage.dart
import 'package:shared_preferences/shared_preferences.dart';
import '../../features/auth/domain/entities/auth_tokens.dart';

class TokenStorage {
  static const _accessTokenKey = 'access_token';
  static const _refreshTokenKey = 'refresh_token';

  Future<void> saveTokens(AuthTokens tokens) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_accessTokenKey, tokens.accessToken);
    await prefs.setString(_refreshTokenKey, tokens.refreshToken);
  }

  Future<AuthTokens?> loadTokens() async {
    final prefs = await SharedPreferences.getInstance();
    final accessToken = prefs.getString(_accessTokenKey);
    final refreshToken = prefs.getString(_refreshTokenKey);

    if (accessToken == null || refreshToken == null) return null;
    return AuthTokens(accessToken: accessToken, refreshToken: refreshToken);
  }

  Future<void> clearTokens() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_accessTokenKey);
    await prefs.remove(_refreshTokenKey);
  }
}
